File: Top_100_Liked_Questions/448._Find_All_Numbers_Disappeared_in_an_Array/solution.py
class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        # O(1) space solution
        # use the number as index of the list and set the value of the index as negative
        for n in nums:
            if nums[abs(n)-1] > 0:
                nums[abs(n)-1] *= -1
        
        ans = []
        for i,v in enumerate(nums):
            if v > 0:
                ans.append(i+1)
                
        return ans
            
            
    def findDisappearedNumbersHashMap(self, nums: List[int]) -> List[int]:
        mem = [False] * len(nums)
        ans = []
        
        for n in nums:
            mem[n-1] = True
            
        for i,v in enumerate(mem):
            if v == 0:
                ans.append(i+1)
        
        return ans

File: Top_100_Liked_Questions/448._Find_All_Numbers_Disappeared_in_an_Array/solution.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var findDisappearedNumbers = function(nums) {
    // const space solution
    for (const n of nums) {
        if (nums[Math.abs(n)-1] > 0) {
            nums[Math.abs(n)-1] *= -1;
        }
    }
    
    const ans = [];
    for (const [i,v] of nums.entries()) {
        if (v > 0) {
            ans.push(i+1);
        }
    }
    
    return ans;
}

var findDisappearedNumbersHashMap = function(nums) {
    const mem = new Array(nums.lengath).fill(0);
    
    for (const n of nums) {
        mem[n-1] = 1;
    }
    
    const ans = [];
    for (const [i,v] of mem.entries()) {
        if (v === 0) {
            ans.push(i+1);
        }
    }
    
    return ans;
};